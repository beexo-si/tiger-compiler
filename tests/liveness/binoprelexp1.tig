/*TIGER_ARGS=-code -flow*/
let
    var s:=2
    var x:=34
    var y:= s=x
in
    y := if y then 1 else 2;
    y
end
