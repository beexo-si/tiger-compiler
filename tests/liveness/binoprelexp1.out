bien!
;;--FRAME--L0__tigermain_0:
LABEL: L7
MOVE: movq %'s0, %'d0 D:T5 S:rbx
MOVE: movq %'s0, %'d0 D:T6 S:r12
MOVE: movq %'s0, %'d0 D:T7 S:r13
MOVE: movq %'s0, %'d0 D:T8 S:r14
MOVE: movq %'s0, %'d0 D:T9 S:r15
MOVE: movq %'s0, %'d0 D:T10 S:rbp
OPER: addq $-8, %'d0 D:[T10] S:[T10]
OPER: movq %'s0, (%'s1) D:[] S:[rdi,T10]
OPER: movq $2, %'d0 D:[T11] S:[]
MOVE: movq %'s0, %'d0 D:T0 S:T11
OPER: movq $34, %'d0 D:[T12] S:[]
MOVE: movq %'s0, %'d0 D:T1 S:T12
OPER: movq $1, %'d0 D:[T13] S:[]
MOVE: movq %'s0, %'d0 D:T3 S:T13
OPER: cmpq %'s0, %'s1 D:[] S:[T0,T1]
OPER: je 'j0 D:[] S:[] J:[L1,L2]
LABEL: L2
OPER: movq $0, %'d0 D:[T14] S:[]
MOVE: movq %'s0, %'d0 D:T3 S:T14
LABEL: L1
MOVE: movq %'s0, %'d0 D:T2 S:T3
OPER: movq $0, %'d0 D:[T15] S:[]
OPER: cmpq %'s0, %'s1 D:[] S:[T2,T15]
OPER: jne 'j0 D:[] S:[] J:[L3,L4]
LABEL: L4
OPER: movq $2, %'d0 D:[T16] S:[]
MOVE: movq %'s0, %'d0 D:T4 S:T16
LABEL: L5
MOVE: movq %'s0, %'d0 D:T2 S:T4
MOVE: movq %'s0, %'d0 D:rax S:T2
MOVE: movq %'s0, %'d0 D:rbx S:T5
MOVE: movq %'s0, %'d0 D:r12 S:T6
MOVE: movq %'s0, %'d0 D:r13 S:T7
MOVE: movq %'s0, %'d0 D:r14 S:T8
MOVE: movq %'s0, %'d0 D:r15 S:T9
OPER: jmp 'j0 D:[] S:[] J:[L6]
LABEL: L3
OPER: movq $1, %'d0 D:[T17] S:[]
MOVE: movq %'s0, %'d0 D:T4 S:T17
OPER: jmp 'j0 D:[] S:[] J:[L5]
LABEL: L6
OPER:  D:[] S:[rax,rsp,rbp,rbx,r12,r13,r14,r15]
;;-END-FRAME-:
;;--FLOW--L0__tigermain_0:
def:
	n0(LABEL: L7): 
	n1(MOVE: movq %'s0, %'d0 D:T5 S:rbx): T5
	n2(MOVE: movq %'s0, %'d0 D:T6 S:r12): T6
	n3(MOVE: movq %'s0, %'d0 D:T7 S:r13): T7
	n4(MOVE: movq %'s0, %'d0 D:T8 S:r14): T8
	n5(MOVE: movq %'s0, %'d0 D:T9 S:r15): T9
	n6(MOVE: movq %'s0, %'d0 D:T10 S:rbp): T10
	n7(OPER: addq $-8, %'d0 D:[T10] S:[T10]): T10
	n8(OPER: movq %'s0, (%'s1) D:[] S:[rdi,T10]): 
	n9(OPER: movq $2, %'d0 D:[T11] S:[]): T11
	n10(MOVE: movq %'s0, %'d0 D:T0 S:T11): T0
	n11(OPER: movq $34, %'d0 D:[T12] S:[]): T12
	n12(MOVE: movq %'s0, %'d0 D:T1 S:T12): T1
	n13(OPER: movq $1, %'d0 D:[T13] S:[]): T13
	n14(MOVE: movq %'s0, %'d0 D:T3 S:T13): T3
	n15(OPER: cmpq %'s0, %'s1 D:[] S:[T0,T1]): 
	n16(OPER: je 'j0 D:[] S:[] J:[L1,L2]): 
	n17(LABEL: L2): 
	n18(OPER: movq $0, %'d0 D:[T14] S:[]): T14
	n19(MOVE: movq %'s0, %'d0 D:T3 S:T14): T3
	n20(LABEL: L1): 
	n21(MOVE: movq %'s0, %'d0 D:T2 S:T3): T2
	n22(OPER: movq $0, %'d0 D:[T15] S:[]): T15
	n23(OPER: cmpq %'s0, %'s1 D:[] S:[T2,T15]): 
	n24(OPER: jne 'j0 D:[] S:[] J:[L3,L4]): 
	n25(LABEL: L4): 
	n26(OPER: movq $2, %'d0 D:[T16] S:[]): T16
	n27(MOVE: movq %'s0, %'d0 D:T4 S:T16): T4
	n28(LABEL: L5): 
	n29(MOVE: movq %'s0, %'d0 D:T2 S:T4): T2
	n30(MOVE: movq %'s0, %'d0 D:rax S:T2): rax
	n31(MOVE: movq %'s0, %'d0 D:rbx S:T5): rbx
	n32(MOVE: movq %'s0, %'d0 D:r12 S:T6): r12
	n33(MOVE: movq %'s0, %'d0 D:r13 S:T7): r13
	n34(MOVE: movq %'s0, %'d0 D:r14 S:T8): r14
	n35(MOVE: movq %'s0, %'d0 D:r15 S:T9): r15
	n36(OPER: jmp 'j0 D:[] S:[] J:[L6]): 
	n37(LABEL: L3): 
	n38(OPER: movq $1, %'d0 D:[T17] S:[]): T17
	n39(MOVE: movq %'s0, %'d0 D:T4 S:T17): T4
	n40(OPER: jmp 'j0 D:[] S:[] J:[L5]): 
	n41(LABEL: L6): 
	n42(OPER:  D:[] S:[rax,rsp,rbp,rbx,r12,r13,r14,r15]): 
use:
	n0(LABEL: L7): 
	n1(MOVE: movq %'s0, %'d0 D:T5 S:rbx): rbx
	n2(MOVE: movq %'s0, %'d0 D:T6 S:r12): r12
	n3(MOVE: movq %'s0, %'d0 D:T7 S:r13): r13
	n4(MOVE: movq %'s0, %'d0 D:T8 S:r14): r14
	n5(MOVE: movq %'s0, %'d0 D:T9 S:r15): r15
	n6(MOVE: movq %'s0, %'d0 D:T10 S:rbp): rbp
	n7(OPER: addq $-8, %'d0 D:[T10] S:[T10]): T10
	n8(OPER: movq %'s0, (%'s1) D:[] S:[rdi,T10]): T10, rdi
	n9(OPER: movq $2, %'d0 D:[T11] S:[]): 
	n10(MOVE: movq %'s0, %'d0 D:T0 S:T11): T11
	n11(OPER: movq $34, %'d0 D:[T12] S:[]): 
	n12(MOVE: movq %'s0, %'d0 D:T1 S:T12): T12
	n13(OPER: movq $1, %'d0 D:[T13] S:[]): 
	n14(MOVE: movq %'s0, %'d0 D:T3 S:T13): T13
	n15(OPER: cmpq %'s0, %'s1 D:[] S:[T0,T1]): T0, T1
	n16(OPER: je 'j0 D:[] S:[] J:[L1,L2]): 
	n17(LABEL: L2): 
	n18(OPER: movq $0, %'d0 D:[T14] S:[]): 
	n19(MOVE: movq %'s0, %'d0 D:T3 S:T14): T14
	n20(LABEL: L1): 
	n21(MOVE: movq %'s0, %'d0 D:T2 S:T3): T3
	n22(OPER: movq $0, %'d0 D:[T15] S:[]): 
	n23(OPER: cmpq %'s0, %'s1 D:[] S:[T2,T15]): T15, T2
	n24(OPER: jne 'j0 D:[] S:[] J:[L3,L4]): 
	n25(LABEL: L4): 
	n26(OPER: movq $2, %'d0 D:[T16] S:[]): 
	n27(MOVE: movq %'s0, %'d0 D:T4 S:T16): T16
	n28(LABEL: L5): 
	n29(MOVE: movq %'s0, %'d0 D:T2 S:T4): T4
	n30(MOVE: movq %'s0, %'d0 D:rax S:T2): T2
	n31(MOVE: movq %'s0, %'d0 D:rbx S:T5): T5
	n32(MOVE: movq %'s0, %'d0 D:r12 S:T6): T6
	n33(MOVE: movq %'s0, %'d0 D:r13 S:T7): T7
	n34(MOVE: movq %'s0, %'d0 D:r14 S:T8): T8
	n35(MOVE: movq %'s0, %'d0 D:r15 S:T9): T9
	n36(OPER: jmp 'j0 D:[] S:[] J:[L6]): 
	n37(LABEL: L3): 
	n38(OPER: movq $1, %'d0 D:[T17] S:[]): 
	n39(MOVE: movq %'s0, %'d0 D:T4 S:T17): T17
	n40(OPER: jmp 'j0 D:[] S:[] J:[L5]): 
	n41(LABEL: L6): 
	n42(OPER:  D:[] S:[rax,rsp,rbp,rbx,r12,r13,r14,r15]): r12, r13, r14, r15, rax, rbp, rbx, rsp
ismove:
	n0(LABEL: L7): false
	n1(MOVE: movq %'s0, %'d0 D:T5 S:rbx): true
	n2(MOVE: movq %'s0, %'d0 D:T6 S:r12): true
	n3(MOVE: movq %'s0, %'d0 D:T7 S:r13): true
	n4(MOVE: movq %'s0, %'d0 D:T8 S:r14): true
	n5(MOVE: movq %'s0, %'d0 D:T9 S:r15): true
	n6(MOVE: movq %'s0, %'d0 D:T10 S:rbp): true
	n7(OPER: addq $-8, %'d0 D:[T10] S:[T10]): false
	n8(OPER: movq %'s0, (%'s1) D:[] S:[rdi,T10]): false
	n9(OPER: movq $2, %'d0 D:[T11] S:[]): false
	n10(MOVE: movq %'s0, %'d0 D:T0 S:T11): true
	n11(OPER: movq $34, %'d0 D:[T12] S:[]): false
	n12(MOVE: movq %'s0, %'d0 D:T1 S:T12): true
	n13(OPER: movq $1, %'d0 D:[T13] S:[]): false
	n14(MOVE: movq %'s0, %'d0 D:T3 S:T13): true
	n15(OPER: cmpq %'s0, %'s1 D:[] S:[T0,T1]): false
	n16(OPER: je 'j0 D:[] S:[] J:[L1,L2]): false
	n17(LABEL: L2): false
	n18(OPER: movq $0, %'d0 D:[T14] S:[]): false
	n19(MOVE: movq %'s0, %'d0 D:T3 S:T14): true
	n20(LABEL: L1): false
	n21(MOVE: movq %'s0, %'d0 D:T2 S:T3): true
	n22(OPER: movq $0, %'d0 D:[T15] S:[]): false
	n23(OPER: cmpq %'s0, %'s1 D:[] S:[T2,T15]): false
	n24(OPER: jne 'j0 D:[] S:[] J:[L3,L4]): false
	n25(LABEL: L4): false
	n26(OPER: movq $2, %'d0 D:[T16] S:[]): false
	n27(MOVE: movq %'s0, %'d0 D:T4 S:T16): true
	n28(LABEL: L5): false
	n29(MOVE: movq %'s0, %'d0 D:T2 S:T4): true
	n30(MOVE: movq %'s0, %'d0 D:rax S:T2): true
	n31(MOVE: movq %'s0, %'d0 D:rbx S:T5): true
	n32(MOVE: movq %'s0, %'d0 D:r12 S:T6): true
	n33(MOVE: movq %'s0, %'d0 D:r13 S:T7): true
	n34(MOVE: movq %'s0, %'d0 D:r14 S:T8): true
	n35(MOVE: movq %'s0, %'d0 D:r15 S:T9): true
	n36(OPER: jmp 'j0 D:[] S:[] J:[L6]): false
	n37(LABEL: L3): false
	n38(OPER: movq $1, %'d0 D:[T17] S:[]): false
	n39(MOVE: movq %'s0, %'d0 D:T4 S:T17): true
	n40(OPER: jmp 'j0 D:[] S:[] J:[L5]): false
	n41(LABEL: L6): false
	n42(OPER:  D:[] S:[rax,rsp,rbp,rbx,r12,r13,r14,r15]): false
succesors:
	n0(LABEL: L7): n1
	n1(MOVE: movq %'s0, %'d0 D:T5 S:rbx): n2
	n2(MOVE: movq %'s0, %'d0 D:T6 S:r12): n3
	n3(MOVE: movq %'s0, %'d0 D:T7 S:r13): n4
	n4(MOVE: movq %'s0, %'d0 D:T8 S:r14): n5
	n5(MOVE: movq %'s0, %'d0 D:T9 S:r15): n6
	n6(MOVE: movq %'s0, %'d0 D:T10 S:rbp): n7
	n7(OPER: addq $-8, %'d0 D:[T10] S:[T10]): n8
	n8(OPER: movq %'s0, (%'s1) D:[] S:[rdi,T10]): n9
	n9(OPER: movq $2, %'d0 D:[T11] S:[]): n10
	n10(MOVE: movq %'s0, %'d0 D:T0 S:T11): n11
	n11(OPER: movq $34, %'d0 D:[T12] S:[]): n12
	n12(MOVE: movq %'s0, %'d0 D:T1 S:T12): n13
	n13(OPER: movq $1, %'d0 D:[T13] S:[]): n14
	n14(MOVE: movq %'s0, %'d0 D:T3 S:T13): n15
	n15(OPER: cmpq %'s0, %'s1 D:[] S:[T0,T1]): n16
	n16(OPER: je 'j0 D:[] S:[] J:[L1,L2]): n17, n20
	n17(LABEL: L2): n18
	n18(OPER: movq $0, %'d0 D:[T14] S:[]): n19
	n19(MOVE: movq %'s0, %'d0 D:T3 S:T14): n20
	n20(LABEL: L1): n21
	n21(MOVE: movq %'s0, %'d0 D:T2 S:T3): n22
	n22(OPER: movq $0, %'d0 D:[T15] S:[]): n23
	n23(OPER: cmpq %'s0, %'s1 D:[] S:[T2,T15]): n24
	n24(OPER: jne 'j0 D:[] S:[] J:[L3,L4]): n25, n37
	n25(LABEL: L4): n26
	n26(OPER: movq $2, %'d0 D:[T16] S:[]): n27
	n27(MOVE: movq %'s0, %'d0 D:T4 S:T16): n28
	n28(LABEL: L5): n29
	n29(MOVE: movq %'s0, %'d0 D:T2 S:T4): n30
	n30(MOVE: movq %'s0, %'d0 D:rax S:T2): n31
	n31(MOVE: movq %'s0, %'d0 D:rbx S:T5): n32
	n32(MOVE: movq %'s0, %'d0 D:r12 S:T6): n33
	n33(MOVE: movq %'s0, %'d0 D:r13 S:T7): n34
	n34(MOVE: movq %'s0, %'d0 D:r14 S:T8): n35
	n35(MOVE: movq %'s0, %'d0 D:r15 S:T9): n36
	n36(OPER: jmp 'j0 D:[] S:[] J:[L6]): n41
	n37(LABEL: L3): n38
	n38(OPER: movq $1, %'d0 D:[T17] S:[]): n39
	n39(MOVE: movq %'s0, %'d0 D:T4 S:T17): n40
	n40(OPER: jmp 'j0 D:[] S:[] J:[L5]): n28
	n41(LABEL: L6): n42
	n42(OPER:  D:[] S:[rax,rsp,rbp,rbx,r12,r13,r14,r15]): 
;;-END-FLOW-:
yes!!
